src/rcv/septentrio.c

```diff
@@ -119,802 +337,2401 @@ static uint8_t sig_tbl[SBF_MAXSIG+1][2]={ /* system, obs-code */
+        raw->nav.sbssat.sat[j].fcorr.prc = R4(p+12+i*sbLength+4);
+
+        dt = timediff(raw->nav.sbssat.sat[j].fcorr.t0, t0_old);
+        if (t0_old.time==0 || dt<=0.0 || 18.0<dt || raw->nav.sbssat.sat[j].fcorr.ai==0) {
+            raw->nav.sbssat.sat[j].fcorr.rrc = 0.0;
+            raw->nav.sbssat.sat[j].fcorr.dt = 0.0;
+        }
+        else {
+            raw->nav.sbssat.sat[j].fcorr.rrc = (raw->nav.sbssat.sat[j].fcorr.prc-prc_old)/dt;
+            raw->nav.sbssat.sat[j].fcorr.dt = dt;
+        }
+        raw->nav.sbssat.sat[j].fcorr.iodf = iodf;
+    }
+    trace(5, "SBF decode_sbsfast: type=%d iodf=%d\n", U1(p+9), iodf);
+    return 3;
+}
+
+/* decode type 1: prn masks --------------------------------------------------*/
+static int decode_sbsprnmask(raw_t *raw)
+{
+    int i, n, sat, prn;
+    uint8_t *p = (raw->buff)+8;
+
+    trace(4, "SBF decode_sbsprnmask:\n");
+
+    if (raw->len < 17)
+    {
+        trace(1, "SBF decode_sbsprnmask: Block too short\n");
+        return -1;
+    }
+
+    /* get satellite number */
+    prn = U1(p+6);
+    if (prn < 120) return -1;
+    if (prn > 139) return -1;
+
+    raw->nav.sbssat.iodp = U1(p+7);
+    raw->nav.sbssat.nsat = U1(p+8);
+
+    if (raw->outtype) {
+        sprintf(raw->msgtype, "SBF SBAS PRN Mask from PRN=%d", prn);
+    }
+
+    for (n=0; n<raw->nav.sbssat.nsat && n<MAXSAT; n++) {
+       i = U1(p+9+n);
+       if      (i<= 37) sat = satno(SYS_GPS, i);    /*   0- 37: gps */
+       else if (i<= 61) sat = satno(SYS_GLO, i-37); /*  38- 61: glonass */
+       else if (i<=119) sat = 0;                    /*  62-119: future gnss */
+       else if (i<=138) sat = satno(SYS_SBS, i);    /* 120-138: geo/waas */
+       else if (i<=182) sat = 0;                    /* 139-182: reserved */
+       else if (i<=192) sat = satno(SYS_SBS, i+10); /* 183-192: qzss ref [2] */
+       else if (i<=202) sat = satno(SYS_QZS, i);    /* 193-202: qzss ref [2] */
+       else             sat = 0;                    /* 203-   : reserved */
+       raw->nav.sbssat.sat[n].sat = sat;
+    }
+
+    trace(5, "SBF decode_sbsprnmask: nprn=%d iodp=%d\n", n, raw->nav.sbssat.iodp);
+
+    return 3;
+}
+
+/* decode type 6: integrity info ---------------------------------------------*/
+static int decode_sbsintegriy(raw_t *raw)
+{
+    int i, prn;
+    uint8_t *p = (raw->buff)+6;
+    uint8_t iodf[4], udre;
+
+
+    trace(4, "decode_sbsintegriy:\n");
+
+    if (raw->len<71)
+    {
+        trace(1, "SBF decode_sbsintegriy: Block too short\n");
+        return -1;
+    }
+
+    /* get satellite number */
+    prn = U1(p+8);
+    if (prn < 120) return -1;
+    if (prn > 139) return -1;
+
+    if (raw->outtype) {
+        sprintf(raw->msgtype, "SBF SABS Integrity Data from PRN=%d", prn);
+    }
+
+    for (i=0; i<4; i++) {
+        iodf[i] = U1(p+10+i);
+    }
+    for (i=0; i<raw->nav.sbssat.nsat && i<MAXSAT; i++) {
+        if (raw->nav.sbssat.sat[i].fcorr.iodf != iodf[i/13]) continue;
```

Comment:
iodf has only 4 elements but it appears that the index i might range up to MAXSAT-1 and so be OOB. Note clear what the issue is here, might the index into nav.sbssat.sat[i] be different to the index into iodf[], or should iodf[] be expanded, or is the range up to MAXSAT wrong.

Reply:
This is wrong. This was obviously inspired by similar code in decode_sbstype6() in sbas.c, but probably the reference code was already fix.